[FIX] web, website: fix menu auto hide in edit mode

xO-Tx opened the pull request:
Steps to reproduce:

- Go to Website > Add menu items to activate the "auto-hide" (to set the
overflowing menu items in a "+" dropdown).
- Go to "edit" mode > While resizing the window width, the "auto-hide"
menu adaptation is disabled, and overflowing menu items are visible.

The goal of this commit is to fix the behaviour described above (and
potentially, issues that can result from the editor's "unbreakable"
rollbacks on top menu) by preventing the unbreakable mechanism from
detecting header changes and cancelling the auto-hide updates.

Related to opw-3484742

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/134185).

qsm-odoo commented on addons/web/static/src/legacy/js/core/menu.js:
```diff
@@ -61,7 +61,10 @@ export async function initAutoMoreMenu(el, options) {
         clearTimeout(debounce);
-        debounce = setTimeout(_adapt, 250);
+        debounce = setTimeout(() => {
+            el.dispatchEvent(new Event("autoMoreMenu.willAdapt", {bubbles: true}));
+            _adapt();
```
Why is not the `autoMoreMenu.willAdapt` event triggered in the `_adapt` function instead of the debounced version only?

xO-Tx replied:
Just because the issues are mainly caused by the debounced code (as described in the commit message). But Yes, I think it's better to have it in the `_adapt()` to prevent any possible issue caused by the unbreakable rollback... Updated now :+1:

qsm-odoo commented on addons/website/static/src/js/content/menu.js:
```diff
@@ -33,6 +33,10 @@ const BaseAnimatedHeader = animations.Animation.extend({
     start: function () {
+        // Used to prevent the editor's unbreakable protection from restoring
+        // the menu's auto-hide updates in edit mode.
+        this.el.addEventListener("autoMoreMenu.willAdapt", () => this.options.wysiwyg
+            && this.options.wysiwyg.odooEditor.unbreakableStepUnactive());
```
- I see you disable the unbreakable system with `unbreakableStepUnactive`... is it reenabled later automatically?
- Let's use the new syntax for this: `this.options.wysiwyg?.odooEditor.unbreakableStepUnactive()`

xO-Tx replied:
> I see you disable the unbreakable system with unbreakableStepUnactive... is it reenabled later automatically?

Yes, right after the call of `historyStep()`... details:
The [unbreakableStepUnactive() code](https://github.com/odoo/odoo/blob/master/addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js#L1522-L1527C6) will set both the editor's `_toRollback` and `_checkStepUnbreakable` to `false` ... then `_toRollback` will get the `UNBREAKABLE_ROLLBACK_CODE` value once a [change is detected on unbreakable element](https://github.com/odoo/odoo/blob/master/addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js#L873). Also the `_checkStepUnbreakable` flag is [set to true on every call of historyStep()](https://github.com/odoo/odoo/blob/3528f2c3f9d3b30266f3421672986d202c724a76/addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js#L1117) to unable the "unbreakable" check.

> Let's use the new syntax for this: `this.options.wysiwyg?.odooEditor.unbreakableStepUnactive()`

I think it's only allowed on 16.0+, no?  if so, I will add it on fw-port :+1:

qsm-odoo replied:
> I think it's only allowed on 16.0+, no? if so, I will add it on fw-port 👍

I think we only started to use those after 16.0 release but there should be no reason to not be able to in 15.0 (we support the same browsers there). But you are right to be cautious, we can adapt master only to be extra safe.

xO-Tx replied:
Yes, Already tested here on [15.0](https://runbot.odoo.com/runbot/build/50949336)  and it seems that runbot is not OK with it :grimacing: (at least for `<=15.0`).

qsm-odoo reviewed (changes requested):
@xO-Tx 

1. The new editor (with its unbreakable/rollback mechanism) was introduced in 15.0, should not this PR target that version instead?

2. Your commit message is very technical, it's fine since the problem seems complicated but could you also mention clearer *functional* specs for the two problems you are solving? :pray:

xO-Tx commented:
> 1- The new editor (with its unbreakable/rollback mechanism) was introduced in 15.0, should not this PR target that version instead?

@qsm-odoo Yes, good thinking! It seems that the issue is already there since 15.0:

![15 0-editor-unbreakable-issue-mou](https://github.com/odoo/odoo/assets/25032876/2ac57718-198e-4007-ab68-313b3debd215)

The PR targets the 15.0 now :+1:

> Your commit message is very technical, it's fine since the problem seems complicated but could you also mention clearer functional specs for the two problems you are solving? 🙏

Commit message updated! Just let me know if something needs to be adjusted :+1:

qsm-odoo reviewed (changes requested):
@xO-Tx 

- Cannot reproduce the Issue 2 you are explaining, what I am missing? :grimacing: I checked the ticket, the video that was provided explains this but cannot reproduce that either :shrug: I can trust you that this is solved thanks to your commit... but 100% sure?
- Your commit is only partially fixing the first issue: when entering edit mode, the auto-hide feature is not triggered, only when I resize the window. Could you adapt that too? Maybe triggering a resize event once edit mode is started (I thought it was already the case). I guess in 16.0 it works because the iframe is being resized.

xO-Tx commented:
> Cannot reproduce the Issue 2 you are explaining, what I am missing? 😬 I checked the ticket, the video that was provided explains this but cannot reproduce that either 🤷 I can trust you that this is solved thanks to your commit... but 100% sure?

@qsm-odoo I was sure I got the second issue when I was testing the code (on +`16.0` if I'm not mistaken) and that this diff should solve it... But Yes, as you noticed, it was solved unintentionally by this PR: https://github.com/odoo/odoo/pull/131026 (fw-port merged 2 days ago).

If you remember point **[C]** in the technical explanation:

> [C]- When the "Cookies Bar" is enabled: the rollback in [B] will trigger
> a "historyRevert" event which will also trigger_up a
> "widgets_start_request" (see: web_editor > SnippetsMenu > start())
> 
> This will destroy public widgets, including the noBackdropPopupWidget
> of the Cookies Bar, and trigger a window "resize" event (see:
> noBackdropPopupWidget > destroy()).

The code that triggers the `"widgets_start_request"` turned out to be dead and removing it in the PR above broke the flow that leads to the bug here (the "unbreakable rollback" is still guilty though). The bug should be solved now :+1: I also removed everything related to the second issue from this PR.

> Your commit is only partially fixing the first issue: when entering edit mode, the auto-hide feature is not triggered, only when I resize the window. Could you adapt that too? Maybe triggering a resize event once edit mode is started (I thought it was already the case). I guess in 16.0 it works because the iframe is being resized.

Thanks for spotting this! this should happen only before `16.0` , actually, there is already a code that was used to trigger the `"resize"` but I think it will only fires jQuery event handlers :grimacing:, fixed now :+1:

qsm-odoo reviewed (approved):
@robodoo r+

(Note: next time, you can leave the opw mentioned maybe with `Related to opw-XXX` instead of just `opw-XXX`, that way we know where this comes from, but not critical)

For ref: opw-3484742

robodoo commented:
@qsm-odoo you may want to rebuild or fix this PR as it has failed CI.

qsm-odoo commented:
@xO-Tx Can you check about the failed CI? Seems related to menus deletion so... :grimacing:

qsm-odoo commented on addons/web/static/src/legacy/js/core/menu.js:
```diff
@@ -202,6 +203,16 @@ export async function initAutoMoreMenu(el, options) {
         extraItemsToggle.appendChild(extraItemsToggleLink);
         extraItemsToggle.appendChild(dropdownMenu);
         el.insertBefore(extraItemsToggle, target);
+        // TODO Adapt in 16.0: The dropdown menu is closed when clicking inside
+        // the extra menu items. The goal here is to prevent this default
+        // behaviour on "edit" mode to allow correct editing of extra menu
+        // items, mega menu content... This should be simply replaced by the BS5
+        // `autoClose` option.
+        if (window.jQuery && document.body.classList.contains("editor_enable")) {
+            $(extraItemsToggle).on("hide.bs.dropdown", e => {
+                return !e.clickEvent || !e.clickEvent.target.closest(".o_extra_menu_items");
+            });
+        }
```
Obviously, not a fan, but since it's going away in the next version and solves the issue here, let's not overthink this :+1:

qsm-odoo reviewed (approved):
@robodoo r+
